x/twap/logic_test.go

```diff
@@ -521,3 +522,421 @@ func (s *TestSuite) TestPruneRecords() {
+			preSetRecords: []types.TwapRecord{baseRecord},
+			poolId:        baseRecord.PoolId,
+			blockTime:     baseRecord.Time.Add(time.Second),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:  baseRecord.Asset0Denom,
+					quoteDenom: baseRecord.Asset1Denom,
+					overrideSp: sdk.NewDec(2),
+				},
+				{
+					baseDenom:  baseRecord.Asset1Denom,
+					quoteDenom: baseRecord.Asset0Denom,
+					overrideSp: sdk.NewDecWithPrec(2, 1),
+				},
+				{
+					baseDenom:  baseRecord.Asset1Denom,
+					quoteDenom: "extra denom B",
+					overrideSp: sdk.NewDecWithPrec(3, 1),
+				},
+			},
+
+			expectError: types.InvalidRecordCountError{Expected: 3, Actual: 1},
+		},
+		"two-asset; pre-set record at t; updated valid spot price": {
+			preSetRecords: []types.TwapRecord{baseRecord},
+			poolId:        baseRecord.PoolId,
+			blockTime:     baseRecord.Time.Add(time.Second),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:  baseRecord.Asset0Denom,
+					quoteDenom: baseRecord.Asset1Denom,
+					overrideSp: sdk.NewDec(2),
+				},
+				{
+					baseDenom:  baseRecord.Asset1Denom,
+					quoteDenom: baseRecord.Asset0Denom,
+					overrideSp: sdk.NewDecWithPrec(2, 1),
+				},
+			},
+
+			expectedMostRecentRecords: []expectedResults{
+				{
+					spotPriceA: sdk.NewDec(2),
+					spotPriceB: sdk.NewDecWithPrec(2, 1),
+				},
+			},
+			expectedHistoricalRecords: []expectedResults{
+				// The original record.
+				{
+					spotPriceA: baseRecord.P0LastSpotPrice,
+					spotPriceB: baseRecord.P1LastSpotPrice,
+				},
+				// The new record added.
+				{
+					spotPriceA: sdk.NewDec(2),
+					spotPriceB: sdk.NewDecWithPrec(2, 1),
+				},
+			},
+		},
+		"two-asset; pre-set record at t; updated with spot price error in both denom pairs": {
+			preSetRecords: []types.TwapRecord{baseRecord},
+			poolId:        baseRecord.PoolId,
+			blockTime:     baseRecord.Time.Add(time.Second),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:   baseRecord.Asset0Denom,
+					quoteDenom:  baseRecord.Asset1Denom,
+					overrideErr: spError,
+				},
+				{
+					baseDenom:   baseRecord.Asset1Denom,
+					quoteDenom:  baseRecord.Asset0Denom,
+					overrideSp:  sdk.NewDecWithPrec(2, 1),
+					overrideErr: spError,
+				},
+			},
+
+			expectedMostRecentRecords: []expectedResults{
+				{
+					spotPriceA:    sdk.ZeroDec(),
+					spotPriceB:    sdk.NewDecWithPrec(2, 1),
+					lastErrorTime: baseRecord.Time.Add(time.Second), // equals to block time
+				},
+			},
+			expectedHistoricalRecords: []expectedResults{
+				// The original record.
+				{
+					spotPriceA: baseRecord.P0LastSpotPrice,
+					spotPriceB: baseRecord.P1LastSpotPrice,
+				},
+				// The new record added.
+				{
+					spotPriceA:    sdk.ZeroDec(),
+					spotPriceB:    sdk.NewDecWithPrec(2, 1),
+					lastErrorTime: baseRecord.Time.Add(time.Second), // equals to block time
+				},
+			},
+		},
+		"two-asset; pre-set record at t; large spot price in one of the pairs": {
```

Comment:
nit: Maybe instead of "large spot price" we should specify "invalidly large spot price"

Reply:
It is not invalid to have a large spot price. However, we do cap it at a specific maximum